### ISSUE TITLE
ilk type

### ISSUE THREAD

**Issue by glints** [EthereumDeveloperAlliance/flash-loan-open-vault#1]
Hi, thanks for sharing this;
I'm playing around with it, but I can't figure out what this is:

bytes32 _ilk // Maker Ilk Type

what info do I need to provide here?

Cheers,

E

**Comment by kamescg** [comment]
No problem @glints 

The Maker contracts references 32 byte (instead of a string) when creating a new vault. 

For ETH that's `ETH-A` and can be parsed with `ethers`.

**Example**
`const ETHBytes32 = ethers.utils.formatBytes32String('ETH-A')`   

Does that make sense?

**Comment by glints** [comment]
OK,thanks for the quick response; is there a list somewhere? I'm looking to create a USDC vault, would that be something like
ethers.utils.formatBytes32String('USDC-A') ?

**Comment by kamescg** [comment]
Yes - that should work.

Only a couple of asset types exist.

- ETH-A
- USDC-A
- BAT-A
- TBTC-A (Haven't tested this one yet)

**Comment by kamescg** [comment]
I recommend testing your smart contracts on a test network.

You can find all of the essential addresses at https://changelog.makerdao.com/

**Comment by glints** [comment]
Thanks a lot.
Does this mean the flashloan you receive is really USDC-A and not USDC? (I use the [USDC address](https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48) right now) Because on the Aave website they have an [example](https://docs.aave.com/developers/tutorials/performing-a-flash-loan/...-in-your-project) with DAI, but the DAI address they use is the 'real DAI' not a derivative like DAI-A. So that's a bit confusing...

One last thing, sorry for all the questions; I usually test my contracts on kovan or ropsten, but those are pretty simple contracts. These flashloan implementations require a lot of dependencies, and a lot of the time the tokens don't have a testcontract, or one token you need has a testimplementation on Kovan for example, and another token you need has one on ropsten, etc..

What would be the best practice here? Implement your own versions? But even if I implement the tokens myself, I also use curve.fi for example; I can't find a test version; does this mean I should deploy my own version of this as well on testnet or am I missing something obvious?

My whole pipeline is in Python btw.

**Comment by kamescg** [comment]
No - it's still USDC. That's just the reference name.

Testing can be a real issue... Like you said a lot of dependencies exist. And in my experience the easiest way to go about it is testing different features in isolation.

For Curve you can deploy your own version or reach out in the Telegram channel. And they might be able to help setup everything.

**Comment by glints** [comment]
Yes, isolating features and testing this way has also been my approach so far.

Some more questions regarding deployment;

Create FlashLoan MakerDAO Proxy contract:

Step1:
When you call this; https://etherscan.io/address/0x4678f0a6958e4d2bc4f1baf7bc52e8f3564f3fe4#writeContract
Is it the address of the new deployed flashloan contract you have to provide (so call it from the flashloan contract), or the address with which you've deployed the flashloan contract (the owner)? (in the code I see it takes msg.sender as an argument, so I would assume it's the owner that needs to be passed... but not too sure.)

from this contract code I take that this function:

function build(address owner) public returns (DSProxy proxy)
returns the proxy address.

regarding these steps:

Set MakerDAO Proxy Contracts (http://changelog.makerdao.com/)
I take this means calling this function on the flashloan contract (is MakerDao):
setDssProxy(address _dss_proxy) -> is this the same address you need to provide as in what step 1 returns

Deploy GenerateCallData Smart Contract to blockchain. 
after deploying, is this the address that needs to be passed to:

```
function setDataGeneration(address _data_generation) public onlyOwner {
    data_generation = _data_generation;
  }
```
?

Set GenerateCallData Proxy Contracts (http://changelog.makerdao.com/)

I take you mean calling all these functions on the flashloan (is MakerDao) contract:
(and passing the addresses found in https://changelog.makerdao.com/releases/mainnet/latest/contracts.json)

```
function setDataGeneration(address _data_generation) public onlyOwner {
    data_generation = _data_generation;
  }
  function setDssProxy(address _dss_proxy) public onlyOwner {
    dss_proxy = _dss_proxy;
  }
  function setVaultManager(address _vault_manager) public onlyOwner {
    vault_manager = _vault_manager;
  }
  function setJug(address _jug) public onlyOwner {
      dai_jug = _jug;
  }
  function setEthJoin(address _ethJoin) public onlyOwner {
      dai_ethJoin = _ethJoin;
  }
  function setDaiJoin(address _daiJoin) public onlyOwner {
      dai_daiJoin = _daiJoin;
  }
```

Thanks again, and in case I ever pull this off, be sure to DM me an ETH address.

**Comment by glints** [comment]
I figured out most of the variables, deployed everything, but getting stuck at calling openGemVault with a revert;
(I've already implemented Uniswap functionality - swapping eth to usdc without a problem btw in the same contract)
(been reading through MakerDAO docs, but it's a lot to take in, so that's definitely a work in progress)

```
dai_manager = "0x5ef30b9986345249bc32d8928B7ee64DE9435E39"
dai_jug = "0x19c0976f590D67707E62397C87829d896Dc0f1F1"
dai_ethJoin = "0x2F0b23f53734252Bda2277357e97e1517d6B042A"
dai_daiJoin = "0x9759A6Ac90977b93B58547b4A71c78317f391A28"
MCD_JOIN_USDC_A = "0xA191e578a6736167326d05c119CE0c90849E84B7"
MCD_JOIN_TUSD_A = "0x4454aF7C8bb9463203b66C816220D41ED7837f44"
MCD_JOIN_DAI = "0x9759A6Ac90977b93B58547b4A71c78317f391A28"
#PROXY_ACTIONS
vault_manager = "0x82ecD135Dce65Fbc6DbdD0e4237E0AF93FFD5038"
```

I see callVaultManager is payable, do you need to send some eth to it, or does the proxy (generatecalldata) contract need some eth to function?
I've also added an approve before openGemVault on my usdcToken_instance, but I'm not sure which address needs to get approved; the flashloan contract, or the owner of the flashloan contract? I call build with the flashloan contract address as an argument (and thus not with the owner address of the flashloan contract).

When I swap Eth for usdc on uniswap, the usdc gets added to my flashloan contract balance, that's why I'm inclined to think I need to approve the GEM_JOIN address like so (GEM_JOIN == MCD_JOIN_USDC_A here):

```
      IERC20 usdcToken_instance = IERC20(usdcToken);
      DAIinterface daiToken_instance = DAIinterface(daiToken);
      //approve collateral
      // contract now holds the balance
      usdcToken_instance.approve(GEM_JOIN, _amount);

      // breaks here
      openGemVault(GEM_JOIN, ILK, _amount, DAI_BORROW, true);

      //dss_proxy is set in MakerDao and the result of calling build on makerdao proxy registry
      uint256 cdp = getLastCDP(dss_proxy);
      daiToken_instance.approve(GEM_JOIN, _borrow);
      //is there no fee to pay??
      closeGemVault(GEM_JOIN, cdp, _amount);
```
